File: ios/Vocal/Processing/PushRegistrationCoordinator.swift
import Combine
import Foundation

struct PushSession: Equatable, Sendable {
  let uid: String
  let epoch: UInt64
  let installationID: String
}

struct PushBinding: Decodable, Equatable, Sendable {
  let installationID: String
  let active: Bool
  let bindingRevision: Int64
  enum CodingKeys: String, CodingKey {
    case installationID = "installationId"
    case active, bindingRevision
  }
}

struct ProcessingJobHint: Equatable, Sendable {
  let jobID: String
  let eventID: String
  let outcome: String
}

func processingJobHint(_ data: [AnyHashable: Any]) -> ProcessingJobHint? {
  guard data["type"] as? String == "audio_job_outcome",
    let job = data["jobId"] as? String,
    job.range(of: "^[a-fA-F0-9]{24}$", options: .regularExpression) != nil,
    let event = data["eventId"] as? String,
    event.range(of: "^[a-zA-Z0-9_:-]{1,128}$", options: .regularExpression) != nil,
    let outcome = data["outcome"] as? String, ["ready", "failed"].contains(outcome)
  else { return nil }
  return ProcessingJobHint(jobID: job.lowercased(), eventID: event, outcome: outcome)
}

@MainActor protocol PushRegistrationAPI {
  func register(installationID: String, token: String) async throws -> PushBinding
  func deactivate(installationID: String, expectedBindingRevision: Int64) async throws
}

@MainActor final class PushRegistrationAPIClient: PushRegistrationAPI {
  private weak var tokenSource: IDTokenSource?
  private let identityUID: @MainActor () -> String?
  private let transport: AuthHTTPTransport

  init(
    configuration: AuthConfiguration, tokenSource: IDTokenSource,
    identityUID: @escaping @MainActor () -> String?,
    sessionConfiguration: URLSessionConfiguration? = nil
  ) {
    self.tokenSource = tokenSource
    self.identityUID = identityUID
    self.transport = AuthHTTPTransport(
      configuration: configuration,
      sessionConfiguration: sessionConfiguration, rejectRedirects: true)
  }

  func register(installationID: String, token: String) async throws -> PushBinding {
    let path = try route(installationID)
    guard validFCMToken(token) else { throw AuthFailure.invalidInput }
    struct Body: Encodable { let token: String }
    let data = try await send(
      "PUT", path, body: JSONEncoder().encode(Body(token: token)), expectedStatus: 200,
      replay401: true)
    guard let binding = try? JSONDecoder().decode(PushBinding.self, from: data), binding.active,
      binding.installationID.lowercased() == installationID.lowercased(),
      (1...9_007_199_254_740_991).contains(binding.bindingRevision)
    else { throw AuthFailure.malformedResponse }
    return binding
  }

  func deactivate(installationID: String, expectedBindingRevision: Int64) async throws {
    guard (1...9_007_199_254_740_991).contains(expectedBindingRevision) else {
      throw AuthFailure.invalidInput
    }
    struct Body: Encodable { let expectedBindingRevision: Int64 }
    _ = try await send(
      "POST", route(installationID) + "/deactivate",
      body: JSONEncoder().encode(Body(expectedBindingRevision: expectedBindingRevision)),
      expectedStatus: 204, replay401: false)
  }

  private func route(_ id: String) throws -> String {
    guard
      id.range(
        of:
          "^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}$",
        options: .regularExpression) != nil
    else { throw AuthFailure.invalidInput }
    return "/devices/\(id.lowercased())/push"
  }

  private func send(
    _ method: String, _ path: String, body: Data, expectedStatus: Int, replay401: Bool
  ) async throws -> Data {
    guard let tokenSource, let uid = identityUID() else { throw AuthFailure.sessionExpired }
    let session = tokenSource.tokenSession
    func check() throws {
      try Task.checkCancellation()
      guard identityUID() == uid, tokenSource.tokenSession == session else {
        throw CancellationError()
      }
    }
    for attempt in 0...(replay401 ? 1 : 0) {
      try check()
      let token = try await tokenSource.idToken(forceRefresh: attempt == 1)
      try check()
      let (data, response) = try await transport.perform(
        method: method, path: path, body: body, bearer: token)
      try check()
      if response.statusCode == 401, replay401, attempt == 0 { continue }
      guard response.statusCode == expectedStatus else {
        if response.statusCode == 401 { throw AuthFailure.sessionExpired }
        throw AuthFailure.serviceUnavailable
      }
      return data
    }
    throw AuthFailure.sessionExpired
  }
}

private func validFCMToken(_ value: String) -> Bool {
  !value.isEmpty && value.utf8.count <= 4096
    && value.unicodeScalars.allSatisfy { (33...126).contains($0.value) }
}

/// Optional transport is disabled in fixtures/simulator; sanitized local hints remain usable.
@MainActor final class PushRegistrationCoordinator: ObservableObject {
  private struct Registered {
    let session: PushSession
    let token: String
    let binding: PushBinding
  }
  private let api: PushRegistrationAPI
  private let jobs: JobsAPI
  private let session: @MainActor () -> PushSession?
  private let identityUID: @MainActor () -> String?
  private let tokenSource: @MainActor () async -> String?
  private let enabled: Bool
  private var permitted = false
  private var apnsReady = false
  private var token: String?
  private var tokenVersion = 0
  private var generation = 0
  private var registered: Registered?
  private var synchronization: (id: UUID, task: Task<Void, Never>)?
  private var syncAgain = false
  private var forceSync = false
  private var navigation: (id: UUID, hint: ProcessingJobHint, task: Task<Job?, Error>)?
  private var seenRefresh: [String] = []
  private var seenTaps: [String] = []
  @Published private(set) var pendingTap: ProcessingJobHint?
  @Published private(set) var refreshHint: ProcessingJobHint?

  init(
    api: PushRegistrationAPI, jobs: JobsAPI, session: @escaping @MainActor () -> PushSession?,
    identityUID: @escaping @MainActor () -> String?,
    tokenSource: @escaping @MainActor () async -> String?, enabled: Bool
  ) {
    self.api = api
    self.jobs = jobs
    self.session = session
    self.identityUID = identityUID
    self.tokenSource = tokenSource
    self.enabled = enabled
  }

  func permissionChanged(granted: Bool) {
    permitted = granted
    schedule()
  }
  func apnsTokenReady() {
    apnsReady = true
    schedule()
  }
  func apnsRegistrationFailed() {
    apnsReady = false
    synchronization?.task.cancel()
  }
  func receivedFCMToken(_ value: String?) {
    guard let value, validFCMToken(value) else { return }
    token = value
    tokenVersion &+= 1
    schedule()
  }
  func sessionChanged() {
    generation &+= 1
    synchronization?.task.cancel()
    synchronization = nil
    navigation?.task.cancel()
    navigation = nil
    // Keep acknowledged old binding for conditional before-sign-out cleanup.
    schedule(force: true)
  }
  func foreground() { schedule(force: true) }

  func synchronize() async {
    schedule()
    await synchronization?.task.value
  }

  private func schedule(force: Bool = false) {
    guard enabled else { return }
    syncAgain = true
    forceSync = forceSync || force
    guard synchronization == nil else { return }
    let id = UUID()
    let task = Task { @MainActor [weak self] in
      guard let self else { return }
      defer { if self.synchronization?.id == id { self.synchronization = nil } }
      repeat {
        self.syncAgain = false
        let forced = self.forceSync
        self.forceSync = false
        do { try await self.syncOnce(force: forced) } catch is CancellationError { return } catch {
          // Retry optional push on foreground; never log tokens/provider diagnostics.
        }
      } while self.syncAgain && !Task.isCancelled
    }
    synchronization = (id, task)
  }

  private func syncOnce(force: Bool) async throws {
    try Task.checkCancellation()
    guard permitted, apnsReady, let captured = session(), captured.uid == identityUID() else {
      return
    }
    let ticket = generation
    let version = tokenVersion
    let availableToken: String?
    if let token { availableToken = token } else { availableToken = await tokenSource() }
    guard let candidate = availableToken, validFCMToken(candidate) else { return }
    try Task.checkCancellation()
    guard current(captured, ticket), permitted, apnsReady, tokenVersion == version else { return }
    token = candidate
    if !force, let registered, registered.session == captured, registered.token == candidate {
      return
    }
    let binding = try await api.register(installationID: captured.installationID, token: candidate)
    try Task.checkCancellation()
    guard current(captured, ticket) else { return }
    registered = Registered(session: captured, token: candidate, binding: binding)
  }

  func beforeSignOut(uid: String, installationID: String?) async {
    generation &+= 1
    synchronization?.task.cancel()
    synchronization = nil
    syncAgain = false
    navigation?.task.cancel()
    navigation = nil
    pendingTap = nil
    refreshHint = nil
    let old = registered
    registered = nil
    guard enabled, let old, old.session.uid == uid, old.session.installationID == installationID,
      identityUID() == uid
    else { return }
    await performBoundedSignOutCleanup(
      before: { [self] in
        guard self.identityUID() == uid else { return }
        try await self.api.deactivate(
          installationID: old.binding.installationID,
          expectedBindingRevision: old.binding.bindingRevision)
      }, clearIdentity: {}, timeoutNanoseconds: 1_500_000_000)
  }

  func receivedMessage(_ data: [AnyHashable: Any]) {
    if let hint = processingJobHint(data) { receivedMessage(hint) }
  }
  func receivedMessage(_ hint: ProcessingJobHint) {
    guard let captured = session(), captured.uid == identityUID(),
      !seenRefresh.contains(hint.eventID)
    else { return }
    remember(hint.eventID, in: &seenRefresh)
    refreshHint = hint
  }
  func rememberTap(_ data: [AnyHashable: Any]) {
    if let hint = processingJobHint(data) { rememberTap(hint) }
  }
  func rememberTap(_ hint: ProcessingJobHint) {
    guard !seenTaps.contains(hint.eventID) else { return }
    pendingTap = hint
  }

  func resolvePendingTap() async throws -> Job? {
    guard let hint = pendingTap, let captured = session(), captured.uid == identityUID() else {
      return nil
    }
    if let running = navigation, running.hint == hint { return try await running.task.value }
    navigation?.task.cancel()
    let ticket = generation
    let id = UUID()
    let task = Task { @MainActor [weak self] () throws -> Job? in
      guard let self else { return nil }
      let job: Job
      do { job = try await self.jobs.detail(id: hint.jobID) } catch is CancellationError {
        return nil
      } catch JobsFailure.notFound {
        if self.current(captured, ticket), self.pendingTap == hint {
          self.pendingTap = nil
          self.remember(hint.eventID, in: &self.seenTaps)
        }
        return nil
      }
      guard self.current(captured, ticket), self.pendingTap == hint, job.id == hint.jobID else {
        return nil
      }
      self.pendingTap = nil
      self.remember(hint.eventID, in: &self.seenTaps)
      return job
    }
    navigation = (id, hint, task)
    defer { if navigation?.id == id { navigation = nil } }
    return try await task.value
  }

  private func current(_ captured: PushSession, _ ticket: Int) -> Bool {
    generation == ticket && session() == captured && identityUID() == captured.uid
  }
  private func remember(_ event: String, in values: inout [String]) {
    values.append(event)
    if values.count > 128 { values.removeFirst(values.count - 128) }
  }
}

File: android/app/src/main/java/com/hatem/musicmute/processing/PushRegistrationCoordinator.kt
package com.hatem.musicmute.processing

import com.hatem.musicmute.auth.AuthApiClient
import com.hatem.musicmute.auth.AuthFailure
import com.hatem.musicmute.auth.AuthProblem
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job as CoroutineJob
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.*

/** Supply only after successful backend device/session synchronization (not cached offline login). */
data class PushSession(val uid: String, val epoch: Long, val installationId: String)
data class PushBinding(val installationId: String, val active: Boolean, val bindingRevision: Long)
data class ProcessingJobHint(val jobId: String, val eventId: String, val outcome: String)

fun processingJobHint(data: Map<String, String>): ProcessingJobHint? {
    if (data["type"] != "audio_job_outcome") return null
    val job = data["jobId"]?.takeIf { it.matches(Regex("[a-fA-F0-9]{24}")) } ?: return null
    val event = data["eventId"]?.takeIf { it.matches(Regex("[a-zA-Z0-9_:-]{1,128}")) } ?: return null
    val outcome = data["outcome"]?.takeIf { it == "ready" || it == "failed" } ?: return null
    return ProcessingJobHint(job.lowercase(java.util.Locale.ROOT), event, outcome)
}

interface PushRegistrationApi {
    suspend fun register(installationId: String, token: String): PushBinding
    suspend fun deactivate(installationId: String, expectedBindingRevision: Long)
}

class AuthPushRegistrationApi(private val auth: AuthApiClient) : PushRegistrationApi {
    override suspend fun register(installationId: String, token: String): PushBinding {
        validateInstallation(installationId)
        if (!validPushToken(token)) throw AuthFailure(AuthProblem.INVALID_INPUT)
        val body = auth.request("PUT", "/devices/$installationId/push",
            buildJsonObject { put("token", token) }.toString(), expectedStatus = 200)
        return try {
            val objectValue = Json.parseToJsonElement(body).jsonObject
            val id = objectValue["installationId"]?.jsonPrimitive
            val active = objectValue["active"]?.jsonPrimitive
            val revision = objectValue["bindingRevision"]?.jsonPrimitive
            val number = revision?.takeIf { !it.isString }?.longOrNull
            if (id?.isString != true || !id.content.equals(installationId, ignoreCase = true) ||
                active?.isString != false || active.booleanOrNull != true ||
                number == null || number !in 1..9_007_199_254_740_991L) throw IllegalArgumentException()
            PushBinding(id.content, true, number)
        } catch (_: IllegalArgumentException) { throw AuthFailure(AuthProblem.SERVICE_UNAVAILABLE) }
    }

    override suspend fun deactivate(installationId: String, expectedBindingRevision: Long) {
        validateInstallation(installationId)
        require(expectedBindingRevision in 1..9_007_199_254_740_991L)
        auth.request("POST", "/devices/$installationId/push-deactivations",
            buildJsonObject { put("expectedBindingRevision", expectedBindingRevision) }.toString(),
            replaySafe = false, expectedStatus = 204)
    }
}

private fun validateInstallation(value: String) {
    if (!value.matches(Regex("[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}")))
        throw AuthFailure(AuthProblem.INVALID_INPUT)
}
private fun validPushToken(token: String): Boolean = token.length in 1..4096 && token.all { it.code in 33..126 }

/** Tokens and conditional cleanup bindings are memory-only; failures retry on foreground/session sync. */
class PushRegistrationCoordinator(
    private val api: PushRegistrationApi,
    private val jobsApi: JobsApi,
    private val sessionProvider: () -> PushSession?,
    private val identityUid: () -> String?,
    private val tokenSource: suspend () -> String?,
    private val permitted: () -> Boolean,
    private val enabled: Boolean,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate),
) {
    private data class Registered(val session: PushSession, val token: String, val binding: PushBinding)
    private val lock = Any()
    private val tapLock = Mutex()
    private var generation = 0L
    private var token: String? = null
    private var tokenVersion = 0L
    private var registered: Registered? = null
    private var syncJob: CoroutineJob? = null
    private var syncAgain = false
    private var forceSync = false
    private val seenRefresh = LinkedHashSet<String>()
    private val seenTaps = LinkedHashSet<String>()
    private val mutableRefresh = MutableSharedFlow<ProcessingJobHint>(extraBufferCapacity = 1)
    private val mutableTap = MutableStateFlow<ProcessingJobHint?>(null)
    val refreshHints: SharedFlow<ProcessingJobHint> = mutableRefresh.asSharedFlow()
    val pendingTap: StateFlow<ProcessingJobHint?> = mutableTap.asStateFlow()

    fun onToken(value: String) {
        if (!enabled || !validPushToken(value)) return
        synchronized(lock) { token = value; tokenVersion++ }
        scheduleSync()
    }

    fun onSessionChanged() {
        synchronized(lock) {
            generation++
            syncJob?.cancel()
            syncJob = null
            syncAgain = false
            // Preserve the old acknowledged revision until the before-sign-out hook consumes it.
            // A hint received before initial login remains pending; explicit logout clears it below.
        }
        scheduleSync(force = true)
    }

    fun onForeground() = scheduleSync(force = true)

    private fun scheduleSync(force: Boolean = false) {
        if (!enabled) return
        synchronized(lock) {
            syncAgain = true
            forceSync = forceSync || force
            if (syncJob?.isActive == true) return
            syncJob = scope.launch(start = CoroutineStart.LAZY) {
                do {
                    val forced = synchronized(lock) { syncAgain = false; forceSync.also { forceSync = false } }
                    try { syncOnce(forced) }
                    catch (error: CancellationException) { throw error }
                    catch (_: Exception) { /* Optional push retries on foreground; no token/error logging. */ }
                } while (synchronized(lock) { syncAgain })
            }
            syncJob?.start()
        }
    }

    private suspend fun syncOnce(force: Boolean) {
        currentCoroutineContext().ensureActive()
        if (!permitted()) return
        val session = sessionProvider() ?: return
        if (session.uid != identityUid()) return
        validateInstallation(session.installationId)
        val (ticket, version) = synchronized(lock) { generation to tokenVersion }
        val candidate = synchronized(lock) { token } ?: tokenSource()?.takeIf(::validPushToken) ?: return
        synchronized(lock) {
            if (!current(session, ticket) || tokenVersion != version || !permitted()) return
            token = candidate
            if (!force && registered?.let { it.session == session && it.token == candidate } == true) return
        }
        currentCoroutineContext().ensureActive()
        val response = api.register(session.installationId, candidate)
        synchronized(lock) {
            if (!current(session, ticket)) return
            registered = Registered(session, candidate, response)
        }
    }

    /** UI can clear immediately; the SDK identity must remain old until this bounded callback ends. */
    suspend fun beforeSignOut(uid: String, installationId: String?) {
        val (pending, binding) = synchronized(lock) {
            generation++
            syncAgain = false
            mutableTap.value = null
            val oldJob = syncJob
            syncJob = null
            val old = registered?.takeIf { it.session.uid == uid && it.session.installationId == installationId }
            registered = null
            oldJob to old
        }
        pending?.cancel()
        if (!enabled || binding == null || identityUid() != uid) return
        try {
            withTimeoutOrNull(1_500) {
                pending?.cancelAndJoin()
                if (identityUid() == uid) api.deactivate(binding.binding.installationId, binding.binding.bindingRevision)
            }
        } catch (_: Exception) { /* Never persist/replay deactivation with another session's credentials. */ }
    }

    fun onMessage(data: Map<String, String>) {
        if (!enabled) return
        val hint = processingJobHint(data) ?: return
        synchronized(lock) {
            val session = sessionProvider() ?: return
            if (session.uid != identityUid() || !remember(seenRefresh, hint.eventId)) return
            mutableRefresh.tryEmit(hint)
        }
    }

    fun rememberTap(data: Map<String, String>) {
        if (!enabled) return
        val hint = processingJobHint(data) ?: return
        synchronized(lock) { if (hint.eventId !in seenTaps) mutableTap.value = hint }
    }

    /** Fetch first using current credentials; notification data never authorizes or switches accounts. */
    suspend fun resolvePendingTap(): Job? = tapLock.withLock {
        val hint = synchronized(lock) { mutableTap.value } ?: return@withLock null
        val session = sessionProvider() ?: return@withLock null
        val ticket = synchronized(lock) { generation }
        if (identityUid() != session.uid) return@withLock null
        val job = try { jobsApi.detail(hint.jobId) }
        catch (failure: JobsFailure) {
            if (failure.problem != JobsProblem.JOB_NOT_FOUND) throw failure
            synchronized(lock) {
                if (current(session, ticket) && mutableTap.value == hint) {
                    mutableTap.value = null
                    remember(seenTaps, hint.eventId)
                }
            }
            return@withLock null
        }
        synchronized(lock) {
            if (!current(session, ticket) || mutableTap.value != hint || job.id != hint.jobId) return@withLock null
            mutableTap.value = null
            remember(seenTaps, hint.eventId)
            job
        }
    }

    private fun current(session: PushSession, ticket: Long): Boolean =
        generation == ticket && sessionProvider() == session && identityUid() == session.uid

    private fun remember(values: LinkedHashSet<String>, eventId: String): Boolean {
        if (!values.add(eventId)) return false
        if (values.size > 128) values.remove(values.first())
        return true
    }
}
